src/service-api.ts

```diff
@@ -0,0 +1,232 @@
+
+import { Member } from 'graasp';
+
+import { PlanTaskManager } from './plans/task-manager';
+// local
+import common, {
+  changePlan,
+  getCurrentCustomer,
+  getOwnPlan,
+  getPlan,
+  getPlans,
+  setDefaultCard,
+} from './schemas';
+import { StripeTaskManager } from './stripe/task-manager';
+import { SubscriptionTaskManager } from './subscriptions/task-manager';
+import { API_VERSION } from './util/constants';
+
+interface GraaspSubscriptionsOptions {
+  stripeSecretKey: string;
+  // This is called stripeDefaultPlanPriceId, because the customer is linked to the price instead of the plan
+  // This value should looks like this: product_XXXXXXXXXXXXXXXXXXXXXXXX
+  stripeDefaultProductId: string;
+}
+
+const plugin: FastifyPluginAsync<GraaspSubscriptionsOptions> = async (fastify, options) => {
+  const { stripeSecretKey, stripeDefaultProductId } = options;
+  const {
+    members: { taskManager: memberTaskManager },
+    taskRunner: runner,
+  } = fastify;
+
+  // Constants
+  const stripe = new Stripe(stripeSecretKey, { apiVersion: API_VERSION });
+
+  const stipeTaskManager = new StripeTaskManager(stripe);
+
+  const planTaskManger = new PlanTaskManager();
+
+  const subscriptionTaskManager = new SubscriptionTaskManager();
+
+  // schemas
+  fastify.addSchema(common);
+
+  // When a member is created, we subscribe them to the default free plan
+  // This doesn't create a customer in stripe, we only make the link in our database
+  runner.setTaskPostHookHandler<Member>(
+    memberTaskManager.getCreateTaskName(),
+    async (member, _actor, { handler }) => {
+      const t1 = planTaskManger.createGetPlanTask(member, { planId: stripeDefaultProductId });
+      const t2 = subscriptionTaskManager.createCreateDefaultSubscriptionTask(member);
+      t2.getInput = () => ({ defaultPlanId: t1.result.id });
+
+      // If we use the runner, the member is not created when we want to insert the subscription.
+      // A fix would be to modify the runner allowing it to run task in the current transaction.
+      // We cannot use the task runner because the tasks will be run on a new transaction, and create the follwing SQL:
+      // BEGIN TRANSACTION
+      // insert member
+      //    BEGIN TRANSACTION
+      //    insert subscription (member.id)
+      //    END TRANSACTION
+      // END TRANSACTION
+      for (const t of [t1, t2]) {
+        if (t.getInput) Object.assign(t.input, t.getInput());
+        await t.run(handler);
+      }
+    },
+  );
+
+  // get plans
+  fastify.get('/plans', { schema: getPlans }, async ({ member, log }) => {
+    const task = stipeTaskManager.createGetPlansTask(member);
+    return runner.runSingle(task, log);
+  });
+
+  fastify.get<{ Params: { planId: string } }>(
+    '/plans/:planId',
+    { schema: getPlan },
+    async ({ member, params: { planId }, log }) => {
+      const task = stipeTaskManager.createGetPlansTask(member, { product: planId });
+      return (await runner.runSingle(task, log))[0];
```

Comment:
I think you created a getPlanTask, why not use this task instead of the `many` one?

Reply:
I didn't create a getPlanTask for Stripe (only for local db). If we move the data on our end, this task will disappear and the code should look like this :
```typescript
const t1 = planTaskManager.createGetPlan(member, { id: planId});
const t2 = stripeTaskManager.createGetPrices(member, { id: planId});
t2.getResult = () => ({ t1.result, prices: t2.result});
runner.runSingleSequence([t1, t2], log);
```